esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5885,4 +5892,256 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+            }
+
+            Owned<IEspDFUPartLocations> partLocations = createDFUPartLocations();
+            partLocations->setLocationIndex(nextLocationsIndex);
+            partLocations->setLocation(host.str());
+            dfuPartLocations.append(*partLocations.getClear());
+
+            partCopyLocationsIndexes.append(locationsIndexStr.append(nextLocationsIndex).str());
+            locationMap.setValue(host.str(), nextLocationsIndex);
+            nextLocationsIndex++;
+        }
+
+        Owned<IEspDFUPartCopies> partCopies = createDFUPartCopies();
+        partCopies->setPartIndex(partIndex + 1);
+        partCopies->setLocationIndexes(partCopyLocationsIndexes);
+        dfuPartCopies.append(*partCopies.getClear());
+    }
+    return df->numParts();
+}
+
+static const char *securityInfoVersion="1";
+void CWsDfuEx::getFileMeta(IPropertyTree &metaInfo, IDistributedFile &file, IUserDescriptor *user, CFileAccessRole role, const char *expiryTime, const char *keyPairName, IConstDFUFileAccessRequest &req)
+{
+    metaInfo.setProp("version", securityInfoVersion);
+    metaInfo.setProp("logicalFilename", file.queryLogicalName());
+    metaInfo.setProp("jobId", req.getJobId());
+    metaInfo.setProp("accessType", req.getAccessTypeAsString());
+    StringBuffer userStr;
+    if (user)
+        metaInfo.setProp("user", user->getUserName(userStr).str());
+
+    const char *clusterName = req.getCluster(); // can be null
+    Owned<IFileDescriptor> fDesc = file.getFileDescriptor(clusterName);
+
+    /* NB: If file access security is disabled in the environment, or on a per cluster basis
+     * keyPairName will be blank. In that case the meta data is returned in plain format.
+     * NB2: Dafilesrv's would also require file access security to be disabled in that case,
+     * otherwise they will be denied access.
+     * Should be part of the same configuration setup.
+     */
+#ifdef _USE_OPENSSL
+    if (!isEmptyString(keyPairName)) // without it, meta data is not encrypted
+    {
+        metaInfo.setProp("keyPairName", keyPairName);
+
+        Owned<IPropertyTree> secureMetaInfo = createPTreeFromIPT(&metaInfo);
+        secureMetaInfo->setProp("expiryTime", expiryTime);
+
+        extractFilePartInfo(*secureMetaInfo, *fDesc);
+
+        // 1st serialize to JSON
+        StringBuffer secureMetaInfoJson;
+        toJSON(secureMetaInfo, secureMetaInfoJson);
+
+        // 2nd compress
+        MemoryBuffer compressedSecureMetaInfoMb;
+        fastLZCompressToBuffer(compressedSecureMetaInfoMb, secureMetaInfoJson.length(), secureMetaInfoJson.str());
+
+        // 3rd encrypt
+        const char *publicKeyFName = env->getPublicKeyPath(keyPairName);
+        Owned<CLoadedKey> publicKey = loadPublicKeyFromFile(publicKeyFName, nullptr);
+        MemoryBuffer encryptedSecureMetaInfoMb;
+        aesEncryptWithRSAEncryptedKey(encryptedSecureMetaInfoMb, compressedSecureMetaInfoMb.length(), compressedSecureMetaInfoMb.bytes(), *publicKey);
```

Comment:
I don't think this is correct.
You need to encrypt with the public key if you want the information that is encrypted to be hidden from everyone else that might be eavesdropping.  E.g., if you care that spark, or someone listening to the network traffic can tell where the file parts are.  It provides very little security that to prevent someone impersonating it - the public key is relatively easy to crack, especially since the contents of the data being encrypted can probably be reproduced.

What I understood was needed in this situation was for dafilesrv to be able to verify that the request came from a legitimate source.  In that case some data needs to be signed by the private key, and dafilesrv uses a public key to verify that it really came from the source claimed.

I'm also not sure why you are using rsa and aes.  Generally the combination of ras and aes is used to set up a secure connection - with rsa used to mass the aes key, because aes is much more efficient.  I don't understand why you are encrypting with both.